api/lib.go

```diff
@@ -32,13 +33,17 @@ type Cache struct {
 type Querier = types.Querier
 
 func InitCache(dataDir string, supportedFeatures string, cacheSize uint32, instanceMemoryLimit uint32) (Cache, error) {
-	dir := sendSlice([]byte(dataDir))
-	defer freeAfterSend(dir)
-	features := sendSlice([]byte(supportedFeatures))
-	defer freeAfterSend(features)
+	dataDirBytes := []byte(dataDir)
+	supportedFeaturesBytes := []byte(supportedFeatures)
+
+	d := makeView(dataDirBytes)
+	defer runtime.KeepAlive(dataDirBytes)
```

Comment:
Good find. I had to look up https://godoc.org/runtime#KeepAlive

This forces the Go compiler not to free the data before that point, otherwise it could free it anytime after the last reference I guess (in a background thread during the Rust FFI call). Seems like a possible fix to this malloc issue
sed to live forever (what `KeepAlive` suggests).

Not sure if we want `defer` (which might confuse the compiler) and we don't need to ensure those lines are run at all. I would just place the two `KeepAlive` calls at the end of the function, or after the FFI (line 50)

Reply:
The only reason I used `defer` is to have `makeView` and `KeepAlive` close together in code, making it hard to forget about `KeepAlive`. I agree moving it down in the body without `defer` is sufficient and simpler.

I don't have a strong opinion which way is better.

Reply:
> The only reason I used defer is to have makeView and KeepAlive close together in code, making it hard to forget about KeepAlive when variables change

It's a valid reason. I am less sure about the correctness this way (as `KeepAlive` is only a hint to the compiler/optimizer). If you change variable names, the compiler will catch the error anyway. The real issue is adding a new variable here. I would rather err on the side I am more sure of compiler correctness

Reply:
> as KeepAlive is only a hint to the compiler/optimizer

Nope, KeepAlive takes a reference to the object, enforcing the object to live at least until the `KeepAlive(slice)` call. So for correctness, it only matters that it is called after we are done using the slice from Rust.

> The real issue is adding a new variable here.

Jupp.